[comfoair] Initial contribution for comfoair v2 binding

boehan opened the pull request:
This binding is a migration of the v1 comfoair binding. 
Logic is mainly re-used from v1 version with some straightforward changes/additions:
* Temperature and Voltage readings have been implemented using UoM
* Boolean-type values are now implemented as Switch channels

Since, this is my first bigger java project, any suggestions on code quality improvements are highly appreciated.
Also thanks to @HolgerHees and @gieemek for their valuable work on the v1 binding.

JAR-Files are available from (after initial build):
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.comfoair/

[Community thread](https://community.openhab.org/t/comfoair-new-comfoair-binding/92891)

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,2263 @@
+ * @author Holger Hees - Initial Contribution
+ * @author Hans Böhm - Refactoring
+ */
+public enum ComfoAirCommandType {
```
Please add `@NonNullByDefault`

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,2263 @@
+            read_reply_data_pos = new int[] { 0 };
+            read_reply_data_bits = 0x03;
+        }
+    },
```
I really dislike this kind of field initialization; it is easy to read but very hard to add new fields and maintain. Changing all of these to use standard constructor arguments would easily save a thousand lines of code. It is a lot of work but please do it.

boehan replied:
@cpmeister I'm happy to apply your suggestions to improve the code. However, I'm not quite experienced in java development, so could you please provide an example on this to ensure I get you right?

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * Serial port used for communication.
+     */
+    public @Nullable String serialPort;
```
Isn't this required? I wouldn't make this nullable, instead just set it to a dummy value like an empty string. It will get overwritten by the actual configs.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,290 @@
+        return UnDefType.UNDEF;
+    }
+
+    private class AffectedItemsUpdateThread extends Thread {
```
This class isn't used as a Thread at all, instead it is used as a Runnable.
```suggestion
    private class AffectedItemsUpdateThread implements Runnable {
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+                            // interrupted transmissions are merged
+                            Thread.sleep(100);
+                        } catch (InterruptedException e) {
+                            // ignore interruption
```
Interruptions usually mean that the thread should stop what it is doing and exit. It definitely isn't something that should be ignored in a loop. I would suggest throwing an exception instead, possibly the InterruptedException itself.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+                            dumpData(requestBlock), (responseBlock.length > 0 ? dumpData(responseBlock) : "null"));
+
+                } catch (InterruptedException e) {
+                    // ignore interruption
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,290 @@
+                    State value = dataType.convertToState(response, comfoAirCommandType);
+
+                    if (value == null) {
+                        logger.warn("unexpected value for DATA: {}", ComfoAirSerialConnector.dumpData(response));
```
dumpData is an expensive operation, so it shouldn't be executed if the log wouldn't be output anyway.
```suggestion
                        if(logger.isWarnEnabled()){
                              logger.warn("unexpected value for DATA: {}", ComfoAirSerialConnector.dumpData(response));
                        }
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+                }
+
+                byte[] requestBlock = calculateRequest(requestCmd, requestData);
+                logger.trace("send DATA: {}", dumpData(requestBlock));
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+                                && responseBlock[responseBlock.length - 1] == (byte) 0x0f
+                                && (responseBlock[5] & 0xff) == command.getReplyCmd()) {
+
+                            logger.trace("receive RAW DATA: {}", dumpData(responseBlock));
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+                                if (calculateChecksum(_block) == checksum) {
+
+                                    logger.trace("receive CMD: {} DATA: {}",
+                                            String.format("%02x", command.getReplyCmd()), dumpData(replyData));
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+                            }
+
+                            logger.trace("skip CMD: {} DATA: {}", String.format("%02x", command.getReplyCmd()),
+                                    dumpData(cleanedBlock));
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+
+                    Thread.sleep(1000);
+                    logger.debug("Retry cmd. Last call was not successful. Request: {} Response: {}",
+                            dumpData(requestBlock), (responseBlock.length > 0 ? dumpData(responseBlock) : "null"));
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+     * @return successful flag
+     */
+    private boolean send(byte[] request) {
+        logger.trace("send DATA: {}", dumpData(request));
```
see above

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+        System.arraycopy(cleanedBuffer, 0, _block, 0, _block.length);
+
+        return _block;
+    }
```
please remove the excess blank lines

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,652 @@
+
+                } else {
+                    return new int[0];
+                }
```
please remove excess blank lines

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeBoolean.java:
```diff
@@ -0,0 +1,76 @@
+ * @author Holger Hees - Initial Contribution
+ * @author Hans Böhm - Refactoring
+ */
+public class DataTypeBoolean implements ComfoAirDataType {
```
Please add `@NonNullByDefault`

boehan replied:
I left this out intentionally, since adding this would mean a lot of adoptions throughout the code. `convertFromState` method would have to be changed from `int[]` to `Integer[]`, since `int` is not nullable, with according changes to the referring classes. Beyond that I really don't know if annotations like `@Nullable Integer @Nullable []` are an improvement for the code in terms of readability.
So if not mandatory, I would like to avoid touching working code as far as possible. However, if you think that this would valuably improve the code quality, I will try my best to add these annotations.

cpmeister replied:
Arrays are a little different, using `int @Nullable []` should work. But yes, I agree that it is ugly. If you want you can use `@NonNullByDefault({})` on those array fields so that they can be ignored.

boehan replied:
I decided to stick to `int @Nullable []`. In the end it was less work than expected and I have to admit, that the annotations already served their purpose :wink:

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeMessage.java:
```diff
@@ -0,0 +1,106 @@
+ *
+ * @author Holger Hees - Initial Contribution
+ */
+public class DataTypeMessage implements ComfoAirDataType {
```
Please add `@NonNullByDefault`

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeMessage.java:
```diff
@@ -0,0 +1,106 @@
+                if (errorAhi == 0x80) {
+                    errorCode = "A0";
+                }
+                errorCode = "A:" + (convertToCode(errorAhi) + 8);
```
What is the point of this if it gets overwritten anyway?

boehan replied:
This is unchanged from the OH1 binding. The code is from @HolgerHees, but following the protocol documentation there just seems to be an `else` missing. --> fixed

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeNumber.java:
```diff
@@ -0,0 +1,101 @@
+ *
+ * @author Holger Hees - Initial Contribution
+ */
+public class DataTypeNumber implements ComfoAirDataType {
```
Please add `@NonNullByDefault`

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeRPM.java:
```diff
@@ -0,0 +1,72 @@
+                }
+            }
+
+            return new DecimalType((int) (1875000.0 / value));
```
Where did this number come from?

boehan replied:
transferred value is (1875000 / rpm) per protocol

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeRPM.java:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @author Grzegorz Miasko - Initial Contribution
+ */
+public class DataTypeRPM implements ComfoAirDataType {
```
Please add `@NonNullByDefault`

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeTemperature.java:
```diff
@@ -0,0 +1,65 @@
+ * @author Holger Hees - Initial Contribution
+ * @author Hans Böhm - QuantityTypes
+ */
+public class DataTypeTemperature implements ComfoAirDataType {
```
Please add `@NonNullByDefault`

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeVolt.java:
```diff
@@ -0,0 +1,65 @@
+ * @author Grzegorz Miasko - Initial Contribution
+ * @author Hans Böhm - QuantityTypes
+ */
+public class DataTypeVolt implements ComfoAirDataType {
```
Please add `@NonNullByDefault`

boehan commented:
@cpmeister thx for your review (besides all the other work you seem to put into OH development these days, btw). I will try to address your comments as asap.

cpmeister commented on bundles/org.openhab.binding.comfoair/README.md:
```diff
@@ -0,0 +1,889 @@
+| temperatures#ewtTemperature        | Number:Temperature       | EWT Temperature                   | Temperature of geothermal heat exchanger sensor                                             | yes      | true      |
+| temperatures#heaterTemperature     | Number:Temperature       | Heater Temperature                | Temperature of heater sensor                                                                | yes      | true      |
+| temperatures#cookerhoodTemperature | Number:Temperature       | Cookerhood Temperature            | Temperature of cookerhood sensor                                                            | yes      | true      |
+| times#level0Time                   | Number                   | Level 0 duration                  | Operating hours at level 0 (away)                                                           | yes      | true      |
```
`Operating hours` makes it sound like store hours.

boehan replied:
AFAIK "operating hours" is a common (technical) term for the time of operation of appliances. However, I'm open for suggestions.

cpmeister replied:
What about "up time"?

boehan replied:
Yeah, "uptime" sounds reasonable

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommand.java:
```diff
@@ -0,0 +1,133 @@
+    private @Nullable Integer replyCmd;
+    private int[] requestData;
+    private @Nullable Integer requestValue;
+    private @Nullable Integer dataPosition;
```
Can any of these be made final?

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,292 @@
+
+            if (requestCmd != null) {
+                switch (requestCmd) {
+                    case 0x9f:
```
Would it be possible to give these constants names? Or do they not really have names to begin with?

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeBoolean.java:
```diff
@@ -0,0 +1,73 @@
+
+            if (get_reply_data_pos[0] < data.length) {
+                boolean result = (data[get_reply_data_pos[0]] & get_reply_data_bits) == get_reply_data_bits;
+                return (result) ? OnOffType.ON : OnOffType.OFF;
```
```suggestion
                return OnOffType.from(result);
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeMessage.java:
```diff
@@ -0,0 +1,102 @@
+            int errorEA = data[get_reply_data_pos[2]];
+            int errorAhi = data[get_reply_data_pos[3]];
+
+            String errorCode = "";
```
Can you please use StringBuilder here instead of raw String concatenation?

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeNumber.java:
```diff
@@ -0,0 +1,93 @@
+                } else {
+                    return UnDefType.NULL;
+                }
+            }
```
Can you refactor this code so it can be reused?

boehan commented:
I removed the rules from the README since they were just ported over from the v1 binding and completely untested as I'm not using them myself.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,2263 @@
+        }
+    };
+
+    Logger logger = LoggerFactory.getLogger(ComfoAirCommandType.class);
```
Please make all fields in enums private and **final**.

boehan replied:
done

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,308 @@
+        ComfoAirConfiguration config = getConfigAs(ComfoAirConfiguration.class);
+        String serialPort = config.serialPort;
+
+        if (StringUtils.isNotEmpty(serialPort)) {
+            comfoAirConnector = new ComfoAirSerialConnector(serialPortManager, serialPort, BAUDRATE);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+            return;
+        }
```
It is better to arrange your logic in a fail-fast fashion. Also, caching the field as a local variable will allow you to only have to do null checks once. Almost all the null check warnings regarding class fields can be solved by making a local variable to cache the field's value.
```suggestion
        if (StringUtils.isEmpty(serialPort)) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
            return;
        }
        ComfoAirSerialConnector comfoAirConnector = new ComfoAirSerialConnector(serialPortManager, serialPort, BAUDRATE);
        this.comfoAirConnector = comfoAirConnector;
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,862 @@
+            Constants.REQUEST_GET_ANALOGS, Constants.REPLY_GET_ANALOGS, new int[] { 4 }),
+    ANALOG1_VALUE("analog#analog1Value", DataTypeNumber.class, Constants.REQUEST_SET_ANALOGS, 19, 5, new String[] {},
+            Constants.REQUEST_GET_ANALOGS, Constants.REPLY_GET_ANALOGS, new int[] { 5 }),
+    ANALOG2_MIN("analog#analog2Min", DataTypeNumber.class, Constants.REQUEST_SET_ANALOGS, 19, 6, new String[] {},
```
Just like my `new int[0]` constant suggestion, you should create a `new String[0]` constant as well.
```suggestion
    ANALOG2_MIN("analog#analog2Min", DataTypeNumber.class, Constants.REQUEST_SET_ANALOGS, 19, 6, new String[0],
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,862 @@
+    /**
+     * @return data type for this command key
+     */
+    public @Nullable ComfoAirDataType getDataType() {
```
Instead of having to resort to reflection in order to create instances, you could instead create static singlton instances of all your datatype classes so you can just return those singleton instances here. There is no reason you can't reuse your datatype instances since they are stateless.

Although it might be the case that the compiler won't let you use singleton instances in the enum constructor, in that case you can pass a lambda for the constructor instead like `DataTypeBoolean::new`, the enum constructor would then use `Supplier<? extends ComfoAirDataType>` as a parameter.

Also, I'm pretty sure this method should never return null normally, so remove the `@Nullable`. Instead of returning null you should throw some kind of runtime exception, not that you would need to if you use singletons.

```suggestion
    public ComfoAirDataType getDataType() {
```

boehan replied:
Never worked with Singletons before. Could you give an example how this would look like?

boehan replied:
I think I figured out how to use Singletons for the datatype classes. However, since I can't have a `static getInstance()` method on the `ComfoAirDataType` interface, I cannot figure out how to call the method here.
Any advice?

cpmeister replied:
You can initialize your ComfoAirCommandType enum types with the singleton instances.
So 
```
public enum ComfoAirCommandType{

...
EXAMPLE(..., ExampleComfoAirDataType.getInstance(), ...);
...
private final ComfoAirDataType dataType;
...

private ComfoAirCommandType(..., ComfoAirDataType dataType, ...){
    ...
    this.dataType = dataType;
    ...
}
...
public ComfoAirDataType getDataType() {
     return dataType;
}
...
} 
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,862 @@
+                continue;
+            }
+
+            Integer getCmd = entry.read_command == 0 ? null : Integer.valueOf(entry.read_command);
+            Integer replyCmd = Integer.valueOf(entry.read_reply_command);
```
Java automatically box/unboxes primitive types to their object equivalents so you don't need to do it manually.
```suggestion
            int getCmd = entry.read_command == 0 ? null : entry.read_command;
            int replyCmd = entry.read_reply_command;
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,862 @@
+                    ComfoAirCommandType affectedCommandType = ComfoAirCommandType.getCommandTypeByKey(affectedKey);
+
+                    if (affectedCommandType != null) {
+                        Integer getCmd = affectedCommandType.read_command == 0 ? null
+                                : Integer.valueOf(affectedCommandType.read_command);
+                        Integer replyCmd = Integer.valueOf(affectedCommandType.read_reply_command);
+
+                        ComfoAirCommand command = commands.get(replyCmd);
+
+                        if (command == null) {
+                            command = new ComfoAirCommand(affectedKey, getCmd, replyCmd, new int[0], null, null);
+                            commands.put(replyCmd, command);
+                        } else {
+                            command.addKey(affectedKey);
+                        }
```
This logic is used in two places, so it would be better to refactor it into its own method.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,862 @@
+                    if (affectedCommandType != null) {
+                        Integer getCmd = affectedCommandType.read_command == 0 ? null
+                                : Integer.valueOf(affectedCommandType.read_command);
+                        Integer replyCmd = Integer.valueOf(affectedCommandType.read_reply_command);
```
Your other method makes sure that `read_reply_command != 0`. Just making sure that it doesn't need to be the case here.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,862 @@
+        ComfoAirCommandType commandType = ComfoAirCommandType.getCommandTypeByKey(key);
+
+        if (commandType != null) {
+            Integer getCmd = commandType.read_command == 0 ? null : Integer.valueOf(commandType.read_command);
+            Integer replyCmd = Integer.valueOf(commandType.read_reply_command);
+
+            return new ComfoAirCommand(key, getCmd, replyCmd, new int[0], null, null);
```
This logic is used a lot. Should it not be wrapped into the `ComfoAirCommand` constructor?

boehan replied:
done

boehan commented:
> Since you mentioned you wanted some thorough feedback, I decided to give you thorough feedback. grin

Now you start challenging my (still limited) java skills :smile:. I will try to implement your suggestions, if it improves the code. Though, I will need your guidance on some of them.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1596 @@
+	</channel-type>
+
+	<channel-type id="level0_time" advanced="true">
+		<item-type>Number</item-type>
```
I'm not sure if this is really useful here. The device always returns these values in hours, therefore a transformation to smaller units (minutes, seconds) is pretty useless (or even suggesting a wrong level of accuracy). Pretty much the same accounts for the transformation to bigger values (days, wekks, ...) since these do usually not represent a continouos operating state over such a long duration but counting overall operating hours on that state (therefore also the previous 'operating hours' terminology).
I'm okay with changing this to `Number:Time`, if it makes sense within the OH guidelines, however I would be interested what other @openhab/add-ons-maintainers think.

cpmeister replied:
All readonly channels should use units if possible.

boehan replied:
Alright, will do then. I also noticed your comment in [#1470 (comment)](https://github.com/openhab/openhab-core/pull/1470#issuecomment-634393163). With an appropriate duration format, UoM would make definitly sense here.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1596 @@
+	<channel-group-type id="options">
+		<label>Option States</label>
+		<channels>
+			<channel id="isPreheater" typeId="is_preheater" />
+			<channel id="isBypass" typeId="is_bypass" />
+			<channel id="recuType" typeId="recu_type" />
+			<channel id="recuSize" typeId="recu_size" />
+			<channel id="isChimney" typeId="is_chimney" />
+			<channel id="isCookerhood" typeId="is_cookerhood" />
+			<channel id="isHeater" typeId="is_heater" />
+			<channel id="isEnthalpy" typeId="is_enthalpy" />
+			<channel id="isEWT" typeId="is_ewt" />
```
I already thought about this too, but I couldn't figure out, how to properly implement this. I don't want the user having to set these on thing creation, since most users won't even know their correct values. On the other hand, values from the device can only be fetched after successful connection and therefore after creation of the thing. How would I implement this so that these values are set automatically after thing creation (initialization), but can also be updated manually from the thing configuration? Aren't the values then always overwritten with the device defaults after saving the new config (issueing a new initialization)?

cpmeister replied:
Maybe you could make them thing properties instead?

boehan replied:
That would probably be the best option, even if that means that these options cannot be modified by the user. But in my opinion a common user should not mess with these settings anyway and schould be set by the fitter of the appliance.
Another way that came to my mind would be a separate configuration option to set if the configuration should be pulled from the device or if these are user defined.
However, I will implement them as properties for now (BTW is there an option to set labels for properties?)

boehan replied:
done

boehan commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1596 @@
+	<channel-group-type id="menuP9">
+		<label>Menu P9: Option Control States</label>
+		<channels>
+			<channel id="chimneyState" typeId="chimney_state" />
+			<channel id="bypassState" typeId="bypass_state" />
+			<channel id="ewtState" typeId="ewt_state" />
+			<channel id="heaterState" typeId="heater_state" />
+			<channel id="vControlState" typeId="v_control_state" />
+			<channel id="frostState" typeId="frost_state" />
+			<channel id="cookerhoodState" typeId="cookerhood_state" />
+			<channel id="enthalpyState" typeId="enthalpy_state" />
```
Also, thought about that. Am I correct, that if I want to still predefine the channel in the xml file, I have to add all channels initially and remove the not needed ones afterwards?

cpmeister replied:
> Am I correct, that if I want to still predefine the channel in the xml file, I have to add all channels initially and remove the not needed ones afterwards?

That is one way to do it. You could also not specify any of the channels (only specify channel types) and then dynamically add the channels that you need. Either approach works fine so it is up to you which way you want to do it.

boehan replied:
done

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,308 @@
+            String commandKey = channel.getUID().getId();
+            ComfoAirCommand readCommand = ComfoAirCommandType.getReadCommand(commandKey);
+            if (readCommand != null) {
+                State state = sendCommand(readCommand, commandKey);
+
+                updateState(channel.getUID(), state);
```
could you point me to an example?

cpmeister replied:
```suggestion
                scheduler.submit(() -> {
                    State state = sendCommand(readCommand, commandKey);
                    updateState(channel.getUID(), state);
                });
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeTemperature.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public int @Nullable [] convertFromState(State value, ComfoAirCommandType commandType) {
+        if (((DecimalType) value) instanceof DecimalType) {
```
How come `convertToState` returns a `QuantityType` but this method expects a `DecimalType`?

boehan replied:
All `QuantityType` commands ar transformed to `DecimalType` in the [handler](https://github.com/openhab/openhab-addons/pull/7052/files#diff-5ea2270a1fc35f2c479ea7dac0b38fe6R190-R208). Therefore, it should be save, if an item is defined as `Number` instead of `Number:Temperature`.

cpmeister replied:
That is ok...I guess, but think it is a bit misleading when looking at the code. I think it would be better to remove  the `commandToState` and put all that logic into the individual ComfoAirDataType classes.

cpmeister replied:
The `convertToState` and `convertFromState` classes should be symmetrical in behavior.

boehan replied:
Moved the logic to ComfoAirCommandType and removed some redundant code on the Handler.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,875 @@
+     * @return ComfoAirCommandType identified by replyCmd
+     */
+    public static List<ComfoAirCommandType> getCommandTypesByReplyCmd(int replyCmd) {
+        List<ComfoAirCommandType> commands = new ArrayList<ComfoAirCommandType>();
```
```suggestion
        List<ComfoAirCommandType> commands = new ArrayList<>();
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,875 @@
+     */
+    public static Collection<ComfoAirCommand> getReadCommandsByEventTypes(List<String> keys) {
+
+        Map<Integer, ComfoAirCommand> commands = new HashMap<Integer, ComfoAirCommand>();
```
```suggestion
        Map<Integer, ComfoAirCommand> commands = new HashMap<>();
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,875 @@
+     */
+    public static Collection<ComfoAirCommand> getAffectedReadCommands(String key, Set<String> usedKeys) {
+
+        Map<Integer, ComfoAirCommand> commands = new HashMap<Integer, ComfoAirCommand>();
```
```suggestion
        Map<Integer, ComfoAirCommand> commands = new HashMap<>();
```

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,347 @@
+
+    private final Logger logger = LoggerFactory.getLogger(ComfoAirHandler.class);
+    private @Nullable ScheduledFuture<?> poller;
+    private SerialPortManager serialPortManager;
```
```suggestion
    private final SerialPortManager serialPortManager;
```
Also move this field above the non-final fields.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,347 @@
+                        }
+                    }
+                }
+            } catch (final RuntimeException e) {
```
I think I just took this over from an other binding (don't remember which). Should be save to remove this.

cpmeister replied:
please do

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeTemperature.java:
```diff
@@ -0,0 +1,72 @@
+        if (value instanceof QuantityType) {
+            int[] template = commandType.getChangeDataTemplate();
+
+            template[commandType.getChangeDataPos()] = (int) (((QuantityType<?>) value).doubleValue() + 20) * 2;
```
and here

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/datatypes/DataTypeTime.java:
```diff
@@ -0,0 +1,84 @@
+            int[] possibleValues = commandType.getPossibleValues();
+            int position = commandType.getChangeDataPos();
+
+            int intValue = ((QuantityType<?>) value).intValue();
```
Refactored state conversion again to cover this and avoid code duplication

boehan commented:
@cpmeister I think I have worked in all of your review comments. Ready for another review round.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1465 @@
+	</channel-type>
+
+	<channel-type id="RF_value" advanced="true">
+		<item-type>Number</item-type>
```
You should consider changing this to `Number:Dimensionless` so that you can use Percent units.
She same applies to any other channel uses percentages.

boehan replied:
I would rather not do this as these are not indicated as percentage values in the device documentation. Addtionally I rechecked the limits for these channels. Found out that the maximum for the ´min_value´s is defined with 99, and AFAIK this can not be defined for UoM type channels. So I would rather leave these as is.

cpmeister replied:
You can leave as is then.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1465 @@
+	<channel-group-type id="analog">
+		<label>Analog Inputs</label>
+		<channels>
+			<channel id="isAnalog1" typeId="is_analog1" />
+			<channel id="isAnalog2" typeId="is_analog2" />
+			<channel id="isAnalog3" typeId="is_analog3" />
+			<channel id="isAnalog4" typeId="is_analog4" />
+			<channel id="isRF" typeId="is_RF" />
+			<channel id="analog1Mode" typeId="analog1_mode" />
+			<channel id="analog2Mode" typeId="analog2_mode" />
+			<channel id="analog3Mode" typeId="analog3_mode" />
+			<channel id="analog4Mode" typeId="analog4_mode" />
+			<channel id="RFMode" typeId="RF_mode" />
+			<channel id="analog1Negative" typeId="analog1_negative" />
+			<channel id="analog2Negative" typeId="analog2_negative" />
+			<channel id="analog3Negative" typeId="analog3_negative" />
+			<channel id="analog4Negative" typeId="analog4_negative" />
+			<channel id="RFNegative" typeId="RF_negative" />
+			<channel id="analog1Volt" typeId="analog1_volt" />
+			<channel id="analog1Min" typeId="analog1_min" />
+			<channel id="analog1Max" typeId="analog1_max" />
+			<channel id="analog1Value" typeId="analog1_value" />
+			<channel id="analog2Volt" typeId="analog2_volt" />
+			<channel id="analog2Min" typeId="analog2_min" />
+			<channel id="analog2Max" typeId="analog2_max" />
+			<channel id="analog2Value" typeId="analog2_value" />
+			<channel id="analog3Volt" typeId="analog3_volt" />
+			<channel id="analog3Min" typeId="analog3_min" />
+			<channel id="analog3Max" typeId="analog3_max" />
+			<channel id="analog3Value" typeId="analog3_value" />
+			<channel id="analog4Volt" typeId="analog4_volt" />
+			<channel id="analog4Min" typeId="analog4_min" />
+			<channel id="analog4Max" typeId="analog4_max" />
+			<channel id="analog4Value" typeId="analog4_value" />
+			<channel id="RFMin" typeId="RF_min" />
+			<channel id="RFMax" typeId="RF_max" />
+			<channel id="RFValue" typeId="RF_value" />
+			<channel id="analogMode" typeId="analog_mode" />
+		</channels>
+	</channel-group-type>
```
Instead of grouping all your analog channels under a single group you could use multiple groups such that you would have numbered analog groups instead of numbered analog channels.
```suggestion
	<channel-group-type id="analogRf">
	    <label>Analog RF Inputs</label>
		<channels>
			<channel id="isRF" typeId="is_RF" />
			<channel id="RFMode" typeId="RF_mode" />
			<channel id="RFNegative" typeId="RF_negative" />
			<channel id="RFMin" typeId="RF_min" />
			<channel id="RFMax" typeId="RF_max" />
			<channel id="RFValue" typeId="RF_value" />
			<channel id="analogMode" typeId="analog_mode" />
		</channels>
	</channel-group-type>
	<channel-group-type id="analog">
		<label>Analog Inputs</label>
		<channels>
			<channel id="isAnalog" typeId="is_analog" />
			<channel id="analogMode" typeId="analog_mode" />
			<channel id="analogNegative" typeId="analog_negative" />
			<channel id="analogVolt" typeId="analog_volt" />
			<channel id="analogMin" typeId="analog_min" />
			<channel id="analogMax" typeId="analog_max" />
			<channel id="analogValue" typeId="analog_value" />
		</channels>
	</channel-group-type>
```

When in your thingtype definition:
```
		<channel-groups>
...
			<channel-group id="analogRf" typeId="analogRf" />
			<channel-group id="analog1" typeId="analog" />
			<channel-group id="analog2" typeId="analog" />
			<channel-group id="analog3" typeId="analog" />
			<channel-group id="analog4" typeId="analog" />
...
		</channel-groups>
```

This should allow you to cleanup a significant number of channel type definitions. WDYT?

boehan replied:
Good point. Implemented and will push after a short test.

boehan replied:
done

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+                            }
+                        }
+                    }
+                } catch (IOException e) {
```
You might catch an `InterruptedIOException` here, so you want to make sure that interrupt it handled properly.

boehan replied:
Not sure if interrupting the thread is enough or I also have to throw a RuntimeException or somthing. Pls check

boehan replied:
Not sure if interrupting the thread is enough or I also have to throw a RuntimeException or somthing. Pls check

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+                                dumpData(requestBlock), (responseBlock.length > 0 ? dumpData(responseBlock) : "null"));
+                    }
+                } catch (InterruptedException e) {
+                    logger.warn("Transmission was interrupted: {}", e.getMessage());
```
If a thread receives an interrupt it generally means that it should terminate immediately. You shouldn't continue communication requests with it.

boehan replied:
Not sure if interrupting the thread is enough or I also have to throw a RuntimeException or somthing. Pls check

boehan replied:
@cpmeister I assume I also have to return a value here as well (preferably `return ComfoAirCommandType.Constants.EMPTY_INT_ARRAY`) or at least `break` the loop to quit communication requests (same for the other `InterruptedException`s)?

cpmeister replied:
You could either break the loop or just throw an exception. Since any interrupt here wouldn't be caused by your own code I think the best route would be to throw a RuntimeException and let it propagate as high up the stack as possible, thus hopefully exiting as soon as possible.

cpmeister commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+                            // add wait states around reading the stream, so that
+                            // interrupted transmissions are merged
+                            Thread.sleep(100);
+                        } catch (InterruptedException e) {
+                            logger.warn("Transmission was interrupted: {}", e.getMessage());
```
Same thing here.

boehan replied:
Not sure if interrupting the thread is enough or I also have to throw a RuntimeException or somthing. Pls check

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,1024 @@
+
+        if (commandType != null) {
+            ComfoAirDataType dataType = commandType.getDataType();
+            if (dataType instanceof DataTypeBoolean) {
+                value = (OnOffType) command;
+            } else if (dataType instanceof DataTypeNumber || dataType instanceof DataTypeRPM) {
+                value = (DecimalType) command;
+            } else {
+                if (command instanceof QuantityType<?>) {
+                    value = (QuantityType<?>) command;
+                }
+            }
```
Eclipse tells me that I have to cast `command` to a State, but I think `value = (State) command` would suffice?

cpmeister replied:
I forgot that `Command` doesn't implement `State`, so yes you would have to cast using `value = (State) command`.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,1024 @@
+        return null;
+    }
+
+    @SuppressWarnings("null")
```
this would create a warning that `commands.get(replyCmd)` could never be `null` and the null-check not necessary, which is actually wrong if `commands` does not include `replyCmd`.
See also https://community.openhab.org/t/nonnullbydefault-and-hashmap/79845

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+
+                // Fake read request for ccease properties
+                if (requestData.length <= 0 && requestCmd == 0x37) {
+                    requestData = new int[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
```
Removed this if-block since the according `requestCmd` was already removed

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+        }
+        datasum += 173;
+
+        String hexString = Integer.toHexString(datasum);
+        if (hexString.length() > 2) {
+            hexString = hexString.substring(hexString.length() - 2);
+        }
+
+        return (byte) Integer.parseInt(hexString, 16);
```
not my code, but this should lead to the same result
```suggestion
        return (byte) (datasum & 0xFF);
```

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,1024 @@
+    IS_WTW(ComfoAirBindingConstants.CG_INPUTS_PREFIX + ComfoAirBindingConstants.CHANNEL_IS_WTW,
+            DataTypeBoolean.getInstance(), Constants.REQUEST_GET_INPUTS, Constants.REPLY_GET_INPUTS, new int[] { 1 },
+            0x08),
+    IS_BATHROOM2_SWITCH(
+            ComfoAirBindingConstants.CG_INPUTS_PREFIX + ComfoAirBindingConstants.CHANNEL_IS_BATHROOM2_SWITCH,
+            DataTypeBoolean.getInstance(), Constants.REQUEST_GET_INPUTS, Constants.REPLY_GET_INPUTS, new int[] { 1 },
+            0x10),
```
Does this really improve the code style? IMO it is better readable if all CommandTypes use the same definition for their parameters. I can definitely do this if you think it is valueable or common practice (you are the professional her), but I would refrain from reworking most of these again if it is not worth the effort.

cpmeister replied:
If you feel it is too much work you can skip doing it. I just hate seeing redundancy in my own code but I can understand why you wouldn't want to make this particular change since it provides little benefit.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+                            logger.warn("Transmission was interrupted: {}", e.getMessage());
+                        }
+
+                    } while (inputStream != null && inputStream.available() > 0);
```
This doesn't seem to work correctly. At least responses were always caught late (in the next iteration of the surrounding loop)

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,635 @@
+        block[0] = 0x00;
+        block[1] = (byte) command;
+        block[2] = (byte) length;
+        if (requestData.length > 0) {
+            for (int i = 0; i < requestData.length; i++) {
+                block[i + 3] = (byte) requestData[i];
+            }
+        }
```
Does this preserve the cast `(byte) requestData[i]`? Seems this didn't work for me, but could also have been another issue.

cpmeister replied:
You are right, it doesn't handle the cast.

boehan commented:
@cpmeister that was a tough round, but I think I could address most of your comments. Although most of this is not my original code, I'm happy to improve it as far as I am able to and I think it has already gained a lot in readibility. Also helps a lot to gain deeper knowledge of the underlying logic.
However, I'm ready for the next round and hope we get to an end soon, but I'm sure you'll find some more :stuck_out_tongue_winking_eye:

boehan commented:
Decided to reorganize/reduce the Error channels, since they were partly duplicated and the error message is anyway indicating which kind of error it is

boehan commented:
I don't know why the build failed, but it doesn't seem to be an issue with the PR itself (seems that the build actually succeeded)

fwolter commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1177 @@
+				<context>serial-port</context>
+				<description>Serial port that the ComfoAir is connected to</description>
+			</parameter>
+			<parameter name="refreshInterval" type="integer" max="65535" min="10" required="false">
```
```suggestion
			<parameter name="refreshInterval" type="integer" max="65535" min="10" unit="s" required="false">
```

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java:
```diff
@@ -0,0 +1,912 @@
+            Constants.REPLY_GET_ANALOGS, new int[] { 0 }, 0x10);
+
+    private final String key;
+    private final ComfoAirDataType data_type;
```
Mainly, because this was straightly taken over from the v1 binding. Changed this now.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,434 @@
+                    updateState(channel.getUID(), state);
+                });
+            }
+        } catch (IllegalArgumentException e) {
```
Don't remember what I used as a reference here, but after a second look I think it should be safe to remove this.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,592 @@
+                }
+
+                try {
+                    Thread.sleep(1000);
```
I'm still struggling a bit correctly managing threads. I assume I would have to move not only the `comfoAirConnector.open();` ([ComfoAirHandler.java#L110](https://github.com/openhab/openhab-addons/blob/2f2ee96cc76c9d14ff7fe9c90bc7b044810c564e/bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java#L110)), but also the following block with the ThingStatus updates ([ComfoAirHandler.java#L110-L198](https://github.com/openhab/openhab-addons/blob/2f2ee96cc76c9d14ff7fe9c90bc7b044810c564e/bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java#L110-L198))? Would it then be better to move this to a separate method?

fwolter replied:
Yes, you're right, the code following `open()` must be moved to a separate method, too. This new method can be invoked via `scheduler.submit()`, then.

boehan replied:
Done

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,592 @@
+        return sb.toString();
+    }
+
+    private String dumpData(byte[] data) {
+        StringBuilder sb = new StringBuilder();
+        for (byte ch : data) {
+            sb.append(String.format(" %02x", ch));
+        }
+        return sb.toString();
+    }
```
Should I change this? I think for consistency with the above method for int arrays, this should stay as is.

fwolter replied:
You don't need to. It was only a hint.

boehan commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,592 @@
+        this.connected = connected;
+    }
+
+    @Override
+    public void serialEvent(SerialPortEvent event) {
+        try {
+            logger.trace("RXTX library CPU load workaround, sleep forever");
+            Thread.sleep(Long.MAX_VALUE);
+        } catch (InterruptedException e) {
+        }
+    }
```
Thought so as well, but eclipse tells me that I have to implement the method and I didn't find any other reference...

fwolter replied:
It does because you delcared your class with `implements SerialPortEventListener`. When you remove it, the compiler error should be gone.

boehan replied:
:facepalm: of course

boehan commented:
@fwolter thanks for your review. I think I addressed most of your comments, the rest is open for further clarification. Would be nice if we could get this in the 2.5.8 release.

fwolter commented:
> @fwolter thanks for your review. I think I addressed most of your comments, the rest is open for further clarification. Would be nice if we could get this in the 2.5.8 release.

I'm afraid, since this is a new binding it needs to be reviewed by two maintainers. There are many PRs older than yours waiting for another review.

fwolter commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirHandler.java:
```diff
@@ -0,0 +1,434 @@
+        String serialPort = this.config.serialPort;
+
+        if (serialPort.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
```
You might want to give the user a hint what went wrong by passing a message as a third argument.

fwolter commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,590 @@
+                logger.debug("open(): No such Port: {}", serialPortName);
+                setConnected(false);
+            }
+        } catch (PortInUseException e) {
+            logger.debug("open(): Port in Use Exception: {}", e.getMessage(), e);
+            setConnected(false);
+        } catch (UnsupportedCommOperationException e) {
+            logger.debug("open(): Unsupported Comm Operation Exception: {}", e.getMessage(), e);
+            setConnected(false);
+        } catch (IOException e) {
+            logger.debug("open(): IO Exception: {}", e.getMessage(), e);
+            setConnected(false);
```
It seems a bit unreasonable to log the stack trace e.g. when the serial port is in use.

Suggestion: Throw these exceptions and handle them in `ComfoAirHandler`. You can set the thing offline with the exception's message, which will be displayed in the UI, then.

boehan replied:
Looking at some other bindings, I'm not sure what would be the best approach here. From my point of view there are three options:

1. Just rethrow each Exception and catch them in the Handler individually
2. Rethrow them all as a more generic Exception, e.g. an IOException
3. Add a new `ComfoAirException` class and rethrow as such

What would you recommend?

fwolter replied:
I'd say it's a matter of taste. My favorite is 1., because I think it's the most simple approach with respect to code lines. I use multi-tray/catches with the |-syntax to handle the different exception types in the handler.

fwolter commented on bundles/org.openhab.binding.comfoair/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,1177 @@
+	</channel-type>
+
+	<channel-type id="bypass_state" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Bypass State</label>
+		<description>State of the bypass (ON = open / OFF = closed)</description>
+		<category>Switch</category>
+		<state readOnly="true"/>
```
Did you take a look at the item-type `Contact`? It has the two states OPEN and CLOSED.

boehan replied:
Yes thought of that too already. However, I'm not quite sure if this really fits (better). While it would fit in terms of the state, it is not really a contact but flap that is opening/closing. Additionally, this is a self-regulating function of the device, so it could also be seen as ENABLED/DISABLED. Therefore, and to be consistent with the other device option states, I would prefer to keep this as is.

boehan commented:
@fwolter I think I addressed all your comments.

Additionally I adapted the write-only channels (resets) to support command descriptions. Even though, their are still not rendered as buttons in PaperUI as I would have expected (did not test on other UIs by now), I think should be an improvement in general.
Note: Even though the developer docs say, that command descriptions must only be used on String type channels, https://github.com/openhab/openhab-core/pull/1422 suggests that Number types should be supported as well (it's also used this way in other bindings, e.g. [daikinmadoka](https://github.com/openhab/openhab-addons/blob/1bf6976e6a463bdcf99313fda1a5d9d1fb476f8b/bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/resources/ESH-INF/thing/daikinmadoka.xml#L120-L131))

fwolter commented on bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java:
```diff
@@ -0,0 +1,586 @@
+                }
+                setConnected(true);
+            } else {
+                logger.debug("open(): No such Port: {}", serialPortName);
```
Would be good to throw an exception here, too, to let the user know what's wrong via the UI.

Consider whether the connected flag is still needed when using exceptions for error handling. If I see correctly, the flag is always false if you throw an exception, which would be redundant.

boehan replied:
done

boehan commented:
> Your changes look good, but I saw there are still several checkstyle warnings left. You could take a look at `target/code-analysis/report.html`. There are many static fields, which should be final to become constants.

I now fixed most of them. I suppose those about `@NonNullByDefault`annotations on inner classes are false positives.
BTW, are these warnings reported anywhere else as well? Didn't even know about that file...

fwolter commented:
> I now fixed most of them. I suppose those about `@NonNullByDefault`annotations on inner classes are false positives.

Yes, this is a known bug, which is fixed in OH3.

> BTW, are these warnings reported anywhere else as well? Didn't even know about that file...

Only severe issues like missing `@NonNullByDefault` annotations are logged during the maven build.

boehan commented:
@fwolter anything else that needs to be addressed from your POV?

fwolter commented:
We'll see.

boehan commented:
Any changes this could make it into 2.5.9? I know there are a lot of other PRs to be reviewed, but since this is already partly approved it would be nice if this could make it into the distro before the 2.5 feature freeze.
@cpmeister since you already put quite some effort in reviewing this one, could you take another look?

boehan commented:
Cool to have this merged. 
And thank you @cpmeister and @fwolter for your reviews and suggestions. Really helped to improve the binding especially my own programming skills.
